lib/active_merchant/billing/gateways/flex_charge.rb

```diff
@@ -0,0 +1,327 @@
+        ]
+      end
+
+      def phone_from(options)
+        options[:phone] || options.dig(:billing_address, :phone) || options.dig(:billing_address, :phone_number)
+      end
+
+      def fetch_access_token
+        params = { AppKey: @options[:app_key], AppSecret: @options[:app_secret] }
+        url = build_request_url(:authenticate)
+        response = parse(ssl_post(url, params.to_json, headers))
+        @options[:access_token] = response[:accessToken]
+        @options[:expires] = response[:expires]
+      rescue ResponseError => e
+        raise OAuthResponseError.new(e)
+      end
+
+      def ensure_access_token_credential(try_again = true)
+        return if @options[:access_token] && @options[:expires] && ((@options[:expires] || 0) > DateTime.now.strftime('%Q').to_i)
+
+        fetch_access_token
+      rescue ActiveMerchant::OAuthResponseError => e
+        raise e
+      rescue ResponseError => e
+        raise e unless retry?(try_again, e)
+
+        @options.delete(:access_token)
+        ensure_access_token_credential(false)
+      end
+
+      def retry?(try_again, error)
+        try_again && %w(401 404).include?(error.response.code) && @options[:access_token].present?
+      end
+
+      def generate_uuid
+        SecureRandom.uuid
+      end
+
+      def build_request_url(action, id = nil)
+        "#{test? ? test_url : live_url}#{ENDPOINTS_MAPPING[action] % id}"
+      end
+
+      def headers
+        { 'Content-Type' => 'application/json' }.merge(
+          @options[:access_token] ? { 'Authorization' => "Bearer #{@options[:access_token]}" } : {}
+        )
+      end
+
+      def parse(body)
+        JSON.parse(body).with_indifferent_access
+      end
+
+      def commit(action, post, authorization = nil)
+        ensure_access_token_credential
+        response =
+          begin
+            url = build_request_url(action, authorization)
+            add_post_data(post) if action == :tokenize
+            parse(ssl_post(url, post.to_json, headers))
+          rescue ResponseError => e
+            errors = JSON.parse(e.response.body)
+            { status: e.response.code, errors: errors['message'] || errors }
+          end
+        Response.new(
+          success_from(response),
+          message_from(response),
+          response,
+          authorization: authorization_from(response),
+          test: test?,
+          error_code: error_code_from(response)
+        )
+      end
+
+      def success_from(response)
+        return false unless response[:success]
+
+        return true if response.dig(:transaction, :payment_method, :token)
+
+        SUCCESS_MESSAGES.include?(response[:status])
+      end
+
+      def message_from(response)
+        response[:errors].present? ? response[:errors] : response[:status]
+      end
+
+      def authorization_from(response)
+        response[:orderSessionKey]
+      end
+
+      def error_code_from(response)
+        response[:status] unless success_from(response)
```

Comment:
💬 Comment:

Unless this API is weird an the status is not an error code, is better to have a blank error code than an incorrect one in the field.

Reply:
Done!